### PR TITLE
fix and recactor splits and add more tests

### PR DESCRIPTION
**Pull request by xaberus**
This fixes splits and refactors them to have a common code path. I also extended the tests to catch some corner cases and added a rudimentary checking for dangling links in `cmp_iterator(...)`

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -314,158 +350,285 @@ impl<'list, T> CursorMut<'list, T> {
...
-                    LinkedList {
-                        head: Some(current),
-                        tail: old_tail,
-                        len: old_len - self.current_len,
-                    }
-                }).unwrap_or_else(LinkedList::new)
+        use std::mem::replace;
+
+        match self.current {
+            None => replace(self.list, LinkedList::new()),
```
Comment (4e554c4c) [inline src/lib.rs:409]: Should this be the case? I think that if the current element is the empty element then the returned list should be empty.
This is because suppose someone has a cursor to the last element and calls split, this would return the empty list. This should be the same behavior as if one moved the cursor ahead one and called `split_before`.

File: src/lib.rs [hunk 2]
```diff
@@ -314,158 +350,285 @@ impl<'list, T> CursorMut<'list, T> {
...
+            let tail = {
+                let mut c = list.cursor_mut();
+                if let Some(i) = i {
+                    c.move_next();
+                    for _ in 0..i {
+                        c.move_next();
+                    }
+                }
+                println!(" = {:?}", c.current());
+                c.split_before()
```
Comment (4e554c4c) [inline src/lib.rs:610]: How do these test split when the returned list is not compared?

**Review by 4e554c4c — CHANGES_REQUESTED** [review]
Nice work! Thank you. Just a few comments.

File: src/lib.rs (hunk 1 above)
Comment (xaberus) [reply src/lib.rs:409]: @4e554c4c: I think it depends on what we mean by before/after. Maybe it is a good idea to differentiate between before first/after last element... My reasoning went along: when we create a cursor, then `next()` will give the first element, so split should return all of the elements, draining the original list. Of course when we are iterating from behind it makes more sense to return an empty list

File: src/lib.rs (hunk 2 above)
Comment (xaberus) [reply src/lib.rs:610]: Sorry, I did not understand your question... The equality test is performed in lines 616-621, if that is what you meant. Also `cmp_iterator()` now tests the link both ways.

Comment (4e554c4c) [reply src/lib.rs:610]: yes... sorry I missed this somehow

File: src/lib.rs (hunk 1 above)
Comment (4e554c4c) [reply src/lib.rs:409]: fair. This is ambiguous in the rfc, so it should be resolved there.

**Review by 4e554c4c — APPROVED** [review]

**Comment by 4e554c4c** [comment]
Thanks!
